drivers/flash/flash_gd32.c

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright (c) 2022 BrainCo Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT gd_gd32_flash_controller
+
+#include <errno.h>
+#include <kernel.h>
+#include <devicetree.h>
+#include <drivers/flash.h>
+#include <soc.h>
+
+#include "flash_gd32.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(flash_gd32, CONFIG_FLASH_LOG_LEVEL);
+
+struct flash_gd32_data {
+	struct k_sem mutex;
+};
+
+static struct flash_gd32_data flash_data;
+
+static const struct flash_parameters flash_gd32_parameters = {
+	.write_block_size = SOC_NV_FLASH_WRITE_BLOCK_SIZE,
+	.erase_value = 0xff,
+};
+
+static inline void flash_gd32_lock(void)
+{
+#if defined(CONFIG_SOC_SERIES_GD32F403)
+	FMC_CTL0 |= FMC_CTL0_LK;
+#if SOC_NV_FLASH_SIZE > FMC_BANK0_SIZE
+	FMC_CTL1 |= FMC_CTL1_LK;
+#endif
+#else /* !CONFIG_SOC_SERIES_GD32F403 */
+	FMC_CTL |= FMC_CTL_LK;
+#endif
+}
+
+static inline void flash_gd32_unlock(void)
+{
+#if defined(CONFIG_SOC_SERIES_GD32F403)
+	FMC_KEY0 = UNLOCK_KEY0;
+	FMC_KEY0 = UNLOCK_KEY1;
+#if SOC_NV_FLASH_SIZE > FMC_BANK0_SIZE
+	FMC_KEY1 = UNLOCK_KEY0;
+	FMC_KEY1 = UNLOCK_KEY1;
+#endif
+#else /* !CONFIG_SOC_SERIES_GD32F403 */
+	FMC_KEY = UNLOCK_KEY0;
+	FMC_KEY = UNLOCK_KEY1;
+#endif
+}
+
+static bool flash_gd32_valid_range(off_t offset, size_t len)
+{
+	if (offset > SOC_NV_FLASH_SIZE) {
+		return false;
+	}
+
+	if (len && ((offset + (len - 1)) > SOC_NV_FLASH_SIZE)) {
+		return false;
+	}
+
+	return true;
+}
+
+static int flash_gd32_read(const struct device *dev, off_t offset,
+			   void *data, size_t len)
+{
+	if (!flash_gd32_valid_range(offset, len)) {
+		return -EINVAL;
+	}
+
+	memcpy(data, (uint8_t *)SOC_NV_FLASH_ADDR + offset, len);
+
+	return 0;
+}
+
+static int flash_gd32_write(const struct device *dev, off_t offset,
+			    const void *data, size_t len)
+{
+	struct flash_gd32_data *dev_data = dev->data;
+	int ret;
+
+	if (!flash_gd32_valid_range(offset, len)) {
+		return -EINVAL;
+	}
+
+	k_sem_take(&dev_data->mutex, K_FOREVER);
+
+	flash_gd32_unlock();
+
+	ret = flash_gd32_programming(offset, data, len);
+	if (ret < 0) {
+		LOG_ERR("Failed to write data, ret (%u)", ret);
+		goto error;
+	}
+
+error:
+	flash_gd32_lock();
+
+	k_sem_give(&dev_data->mutex);
+
+	return ret;
+}
+
+static int flash_gd32_erase(const struct device *dev, off_t offset,
+			    size_t size)
+{
+	struct flash_gd32_data *dev_data = dev->data;
+	struct flash_pages_info info;
+	uint32_t begin, end;
+	int ret;
+
+	ret = flash_get_page_info_by_offs(dev, offset, &info);
+	if (ret < 0) {
+		return ret;
+	}
+	begin = info.index;
+
+	ret = flash_get_page_info_by_offs(dev, offset + size - 1, &info);
+	if (ret < 0) {
+		return ret;
+	}
+	end = info.index;
+
+	k_sem_take(&dev_data->mutex, K_FOREVER);
+
+	flash_gd32_unlock();
+
+	for ( ; begin <= end; begin++) {
```

Comment:
This is my personal opinion: when you are not able to fill in all `for` expressions just use `while`, looks better.

Reply:
I think `for` means there have an iteration.